subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java

```diff
@@ -275,6 +277,99 @@ public SubscriptionBase apply(final DefaultSubscriptionBase input) {
+                        if (plan.getPlansAllowedInBundle() != -1 && plan.getPlansAllowedInBundle() > 0) {
+                            int existingAddOnsWithSamePlanName = addonUtils.countExistingAddOnsWithSamePlanName(subscriptionWithAddOns, plan.getName());
+                            int currentAddOnsWithSamePlanName = countCurrentAddOnsWithSamePlanName(entitlementWithAddOnsSpecifier.getEntitlementSpecifier(), catalog, plan.getName(), effectiveDate, callContext);
+                            if ((existingAddOnsWithSamePlanName + currentAddOnsWithSamePlanName) > plan.getPlansAllowedInBundle()) {
+                                // a new ADD_ON subscription of the same plan can't be added because it has reached its limit by bundle
+                                throw new SubscriptionBaseApiException(ErrorCode.SUB_CREATE_AO_MAX_PLAN_ALLOWED_BY_BUNDLE, plan.getName());
+                            }
+                        }
+                    }
+
+                    SubscriptionSpecifier subscription = new SubscriptionSpecifier();
+                    subscription.setRealPriceList(plan.getPriceListName());
+                    subscription.setEffectiveDate(effectiveDate);
+                    subscription.setProcessedDate(now);
+                    subscription.setPlan(plan);
+                    subscription.setInitialPhase(spec.getPhaseType());
+                    subscription.setBuilder(new SubscriptionBuilder()
+                                                    .setId(UUIDs.randomUUID())
+                                                    .setBundleId(bundle.getId())
+                                                    .setCategory(plan.getProduct().getCategory())
+                                                    .setBundleStartDate(effectiveDate)
+                                                    .setAlignStartDate(effectiveDate)
+                                                    .setMigrated(entitlementWithAddOnsSpecifier.isMigrated()));
+
+                    subscriptions.add(subscription);
+                }
+                SubscriptionAndAddOnsSpecifier subscriptionAndAddOnsSpecifier = new SubscriptionAndAddOnsSpecifier();
+                subscriptionAndAddOnsSpecifier.setSubscriptionSpecifiers(subscriptions);
+                subscriptionAndAddOns.add(subscriptionAndAddOnsSpecifier);
+            }
+
+            final List<List<DefaultSubscriptionBase>> result = apiService.createPlansWithAddOns(accountId, subscriptionAndAddOns, callContext);
+            return buildSubscriptionBaseList(result);
+
+        } catch (CatalogApiException e) {
+            throw new SubscriptionBaseApiException(e);
+        }
+    }
+
+    private List<SubscriptionBase> buildSubscriptionBaseList(final List<List<DefaultSubscriptionBase>> defaultSubscriptionBaseList) {
+        List<SubscriptionBase> result = new ArrayList<SubscriptionBase>();
```

Comment:
It sucks that we need that :-( It's difficult to say by just reading the diff, if it could be avoided or not.